src/DataReader/proto_reader.rs

```diff
@@ -0,0 +1,273 @@
+/// - `proto_decision.edge` is `None`.
+pub fn proto_decision_to_decision(
+    proto_decision: ProtoDecision,
+    system: &TransitionSystemPtr,
+    components: Vec<Component>,
+) -> Decision {
+    let proto_state: ProtoState = proto_decision.source.unwrap();
+    let state = proto_state_to_state(proto_state, system);
+
+    let proto_edge: ProtoEdge = proto_decision.edge.unwrap();
+    let decided = proto_edge_to_edge(proto_edge, components);
+
+    Decision::new(state, decided)
+}
+
+/// Consumes a [`ProtoState`] and the borrows the [`TransitionsSystemPtr`] it belongs to and returns the corresponding [`State`].
+///
+/// # Panics
+/// If:
+/// - `state.federation` is `None`.
+/// - `state.location_tuple` is `None`.
+pub fn proto_state_to_state(state: ProtoState, system: &TransitionSystemPtr) -> State {
+    let proto_federation: ProtoFederation = state.federation.unwrap();
+    let federation: OwnedFederation =
+        proto_federation_to_owned_federation(proto_federation, system);
+
+    let proto_location_tuple: ProtoLocationTuple = state.location_tuple.unwrap();
+    let location_tuple =
+        proto_location_tuple_to_location_tuple(&proto_location_tuple, system).unwrap();
+
+    State::create(location_tuple, federation)
+}
+
+fn proto_location_tuple_to_location_tuple(
+    location_tuple: &ProtoLocationTuple,
+    system: &TransitionSystemPtr,
+) -> Option<LocationTuple> {
+    let id_looking_for: Vec<LocationID> = location_tuple
+        .locations
+        .iter()
+        .map(|l| LocationID::Simple {
+            location_id: l.id.to_string(),
+            component_id: l
+                .specific_component
+                .as_ref()
+                .map(|c| c.component_name.to_string()),
+        })
+        .collect();
+
+    system
+        .get_all_locations()
+        .into_iter()
+        .map(|tuple| (tuple.id.clone(), tuple))
+        .map(|(id, tuple)| (id.inorder_vec_tranform(), tuple))
+        .find(|(id, _)| id.iter().eq(id_looking_for.iter()))
+        .map(|(_, tuple)| tuple)
+}
+
+fn proto_edge_to_edge(proto_edge: ProtoEdge, components: Vec<Component>) -> Edge {
+    components
+        .into_iter()
+        .map(|c| c.get_edges().to_owned())
+        .reduce(|acc, es| acc.into_iter().chain(es.into_iter()).collect())
+        .unwrap()
+        .into_iter()
+        .find(|e| e.id == proto_edge.id)
+        .unwrap()
+}
+
+fn proto_constraint_to_constraint(
+    proto_constraint: ProtoConstraint,
+    system: &TransitionSystemPtr,
+) -> Constraint {
+    fn determine_index(clock: ProtoComponentClock, system: &TransitionSystemPtr) -> usize {
+        if clock.clock_name == "0" && clock.specific_component.is_none() {
+            0
+        } else {
+            system
+                .clock_name_and_component_to_index(
+                    &clock.clock_name,
+                    &clock.specific_component.unwrap().component_name,
+                )
+                .unwrap()
+        }
+    }
```

Comment:
We need to handle quotient clocks as well somehow as they cannot have a `clock.specific_component`. This probably involves changes to protobuf.

I also dislike the usage of `"0"` as a constant name for the zero clock, i would prefer an enum type e.g. just use `None` and `Some(..)` from `Option<ProtoComponentClock>` as we already receive in proto.